### PR TITLE
Support addons

### LINKED ISSUE #7: Work with Addons?
I'd like to run this codemod on an addon. When I tried, it deletes all the template re-export javascript files e.g. `app/templates/components/my-component.js` and does nothing else.

### PR DESCRIPTION
**Pull request by chriskrycho**
This adds support for migrating addons, by—

- identifying the type of package it's being run in and migrating files from `app` or `addon` accordingly
- finding and removing imports which set the `layout` on a backing class, since the existing import will be *wrong* and there's no point in *updating* it, since it's already bound correctly in an colocated scenario

Resolves #7.

## Details

Before template colocation, addons had to explicitly import a template and set it via the `layout` property or using the `layout` decorator. With template colocation, those can all be dropped: the build pipeline automatically sets it up correctly (though it can also still be set up manually if users so desire, using `setComponentTemplate`, which is what colocation transforms to in the end).

Accordingly, running the codemod against an addon does more than it does in the (default) app case. As with apps, it *does* colocate the template with the backing class. However, it also looks up the imports for the template in question and removes those along with their binding to the `layout` property of a component class or via the `layout` decorator from `ember-decorators`.

Notes on the change:

- Templates shared by multiple components are excluded. In this case, it is up to end users how to list those; they are reported in the same way other skipped templates (like partials) are.

- App components using the `layout` property or decorator are now updated as well. This is very rare, but the codemod can handle it trivially, so it does.

Additional improvements made along the way:

- Add basic JSDoc type annotations for the `Migrator` class to make working with its properties easier (autocomplete etc.).

- Replace `find` methods with getters for repeated lookups where possible (some have to be rerun after other transforms).

- Reuse the values looked up via the utility types.

### CODE REVIEW

File: lib/migrator.js [hunk 1]
```diff
@@ -1,180 +1,270 @@
...
 
-        // Extract '/components/nested1/nested-component.hbs'
-        const filePathFromAppTemplates = filePathFromApp.slice('app/templates/'.length);
+    let templateFilePaths = this.findClassicComponentTemplates()
+      .filter(skipTemplatesUsedAsLayoutName(componentsWithLayoutName, this))
+      .filter(skipTemplatesUsedAsPartial(componentsWithPartial, this));
 
-        // Extract 'components/nested1/nested-component'
-        const classFilePath = filePathFromAppTemplates.slice(1).replace('.hbs', '');
+    templateFilePaths.forEach(transform);
```
Comment (chriskrycho) [inline lib/migrator.js]: This is not quite correct; it needs to run for *class* paths associated with a template which is actually being migrated. Getting that should be straightforward, either using the `classFilePaths` information or simply running a `.map` on the `templateFilePaths` to transform `templates/components` to `components` and `.hbs` to `.js`.

Comment (chriskrycho) [reply lib/migrator.js]: Resolved via 3ea6f6f

**Comment by chriskrycho** [comment]
(This needs a rebase now that #13 has landed. Also needs to be tested exhaustively.)

**Comment by rwjblue** [comment]
@chriskrycho - Where are we at here? Ready for a review pass?

**Comment by chriskrycho** [comment]
@rwjblue yeah, I think it's ready to review, and it *does* work… mostly. I hit one bug when running it on some addons a few weeks ago and haven't had a chance to chase it down the rest of the way. Can report on the details of that on Monday.

**Comment by chriskrycho** [comment]
@rwjblue I have now fixed every known issue with this. I'm sure others may exist, but it's working correctly across our big app's codebase, and there are tests for everything I've hit.

**Pull request merged**